File: synsema/llm/context.py
"""
Synsema LLM Context Builder.

When the language calls the LLM (reason, decide, analyze, generate),
it should send the FULL context, not just the raw data. The LLM needs:

    - What the program is trying to do (intent)
    - What variables are in scope and their values
    - What rules the owner set
    - What the agent has learned (memory)
    - What step of the task we're on (progress)
    - What capabilities are available
    - What trace we're inside of

This turns a generic LLM call into an INFORMED decision.

Without context:
    "Choose between refund and replace given {issue: broken item}"

With context:
    "You are an agent processing customer support tickets.
     Intent: Handle support tickets and resolve issues.
     Current step: evaluate_action (3 of 5).
     Owner rules: max_refund <= 200, prefer replacement over refund.
     Memory: Last 3 similar cases were resolved with replacement.
     Customer: Alice (loyal, 5 previous orders).
     Choose between refund and replace given {issue: broken item}"
"""

from typing import Dict, List, Any, Optional


class LLMContext:
    """
    Collects and formats context for LLM calls.

    Attached to the interpreter and engine, it gathers
    all available context at the moment of an LLM call.
    """

    def __init__(self):
        self.intent: Optional[str] = None
        self.active_trace: Optional[str] = None
        self.current_step: Optional[str] = None
        self.progress_summary: Optional[str] = None
        self.rules: List[Dict] = []
        self.recent_memory: List[Dict] = []
        self.variables: Dict[str, str] = {}
        self.capabilities: List[str] = []
        self.agent_name: Optional[str] = None
        self.custom_context: Dict[str, str] = {}

    def set_intent(self, intent: str):
        self.intent = intent

    def set_trace(self, trace_name: str):
        self.active_trace = trace_name

    def set_progress(self, step: str, summary: str):
        self.current_step = step
        self.progress_summary = summary

    def set_rules(self, rules: List[Dict]):
        self.rules = rules

    def set_memory(self, entries: List[Dict]):
        self.recent_memory = entries

    def set_variables(self, variables: Dict[str, str]):
        self.variables = variables

    def set_capabilities(self, caps: List[str]):
        self.capabilities = caps

    def build_system_prompt(self) -> str:
        """
        Build a system-level prompt with all available context.

        This goes as a system message or preamble before the actual request.
        """
        parts = []

        parts.append("You are the reasoning engine for a Synsema agent.")

        if self.agent_name:
            parts.append(f"Agent name: {self.agent_name}")

        if self.intent:
            parts.append(f"Program intent: {self.intent}")

        if self.current_step:
            step_info = f"Current task step: {self.current_step}"
            if self.progress_summary:
                step_info += f" ({self.progress_summary})"
            parts.append(step_info)

        if self.active_trace:
            parts.append(f"Inside trace block: {self.active_trace}")

        if self.rules:
            rules_text = "Owner rules in effect:\n"
            for r in self.rules:
                level = r.get("level", "")
                desc = r.get("description", "")
                rules_text += f"  [{level}] {desc}\n"
            parts.append(rules_text.rstrip())

        if self.recent_memory:
            mem_text = "Relevant agent memory:\n"
            for m in self.recent_memory[:5]:
                cat = m.get("category", "")
                content = m.get("content", "")
                mem_text += f"  [{cat}] {content}\n"
            parts.append(mem_text.rstrip())

        if self.variables:
            # Filter out builtins and tasks, show only user data
            user_vars = {
                k: v for k, v in self.variables.items()
                if not v.startswith("SynValue(task:") and not v.startswith("builtin:")
                and not v.startswith("task ")
            }
            if user_vars:
                vars_text = "Visible variables:\n"
                for k, v in list(user_vars.items())[:15]:
                    vars_text += f"  {k} = {v}\n"
                parts.append(vars_text.rstrip())

        if self.capabilities:
            parts.append(f"Available capabilities: {', '.join(self.capabilities)}")

        parts.append(
            "Respond concisely and directly. "
            "When choosing between options, respond with ONLY the chosen option. "
            "When analyzing, be structured and actionable. "
            "Respect all owner rules."
        )

        return "\n\n".join(parts)

    def enrich_request_data(self, operation: str, data: Dict) -> Dict:
        """
        Enrich a request's data dict with context.

        This adds context fields to the data so the prompt builder
        can include them.
        """
        enriched = dict(data)
        enriched["_context"] = {
            "intent": self.intent,
            "trace": self.active_trace,
            "step": self.current_step,
            "rules": self.rules,
            "memory": self.recent_memory,
            "agent": self.agent_name,
        }
        return enriched


def build_contextual_prompt(operation: str, data: Dict,
                            context: LLMContext) -> str:
    """
    Build a complete prompt with context for an LLM operation.

    This replaces the generic _build_prompt in providers.
    """
    system = context.build_system_prompt() if context else ""

    if operation == "reason":
        subject = data.get("subject", "")
        user_context = data.get("context", {})
        prompt = f"Reason about: {subject}\n"
        if user_context:
            prompt += "\nProvided context:\n"
            for k, v in user_context.items():
                prompt += f"  {k}: {v}\n"
        prompt += "\nProvide a clear, structured analysis."

    elif operation == "decide":
        options = data.get("options", "")
        given = data.get("given", "")
        prompt = f"Given: {given}\n\n"
        prompt += f"Choose the best option from: {options}\n"
        prompt += "Respond with ONLY the chosen option, nothing else."

    elif operation == "analyze":
        data_str = data.get("data", "")
        objective = data.get("objective", "")
        prompt = f"Data to analyze:\n{data_str}\n\n"
        prompt += f"Objective: {objective}\n"
        prompt += "Provide a concise, actionable analysis."

    elif operation == "generate":
        target = data.get("target", "")
        given = data.get("given", "")
        params = data.get("parameters", {})
        prompt = f"Generate: {target}\n"
        if given:
            prompt += f"\nBased on: {given}\n"
        if params:
            prompt += "\nParameters:\n"
            for k, v in params.items():
                prompt += f"  {k}: {v}\n"

    else:
        prompt = str(data)

    if system:
        return f"{system}\n\n---\n\n{prompt}"
    return prompt

File: rust/crates/synsema-llm/src/context.rs
//! Constructor de contexto para llamadas al LLM. Port de `synsema/llm/context.py`.
//!
//! Cuando el lenguaje llama al LLM (reason/decide/analyze/generate) envía el contexto
//! COMPLETO: intent, variables en scope, reglas del owner, memoria, paso de progreso,
//! capabilities, trace. Convierte una llamada genérica en una decisión INFORMADA.
//! (Módulo de capa 10; no lo ejercita el corpus de conformidad.)

use std::collections::HashMap;

use indexmap::IndexMap;

#[derive(Default)]
pub struct LLMContext {
    pub intent: Option<String>,
    pub active_trace: Option<String>,
    pub current_step: Option<String>,
    pub progress_summary: Option<String>,
    pub rules: Vec<(String, String)>,        // (level, description)
    pub recent_memory: Vec<(String, String)>, // (category, content)
    pub variables: IndexMap<String, String>,
    pub capabilities: Vec<String>,
    pub agent_name: Option<String>,
}

impl LLMContext {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn set_intent(&mut self, intent: &str) {
        self.intent = Some(intent.to_string());
    }
    pub fn set_progress(&mut self, step: &str, summary: &str) {
        self.current_step = Some(step.to_string());
        self.progress_summary = Some(summary.to_string());
    }

    /// Prompt de sistema con todo el contexto disponible.
    pub fn build_system_prompt(&self) -> String {
        let mut parts: Vec<String> = Vec::new();
        parts.push("You are the reasoning engine for a Synsema agent.".to_string());
        if let Some(a) = &self.agent_name {
            parts.push(format!("Agent name: {}", a));
        }
        if let Some(i) = &self.intent {
            parts.push(format!("Program intent: {}", i));
        }
        if let Some(step) = &self.current_step {
            let mut s = format!("Current task step: {}", step);
            if let Some(sum) = &self.progress_summary {
                s += &format!(" ({})", sum);
            }
            parts.push(s);
        }
        if let Some(t) = &self.active_trace {
            parts.push(format!("Inside trace block: {}", t));
        }
        if !self.rules.is_empty() {
            let mut text = String::from("Owner rules in effect:\n");
            for (level, desc) in &self.rules {
                text += &format!("  [{}] {}\n", level, desc);
            }
            parts.push(text.trim_end().to_string());
        }
        if !self.recent_memory.is_empty() {
            let mut text = String::from("Relevant agent memory:\n");
            for (cat, content) in self.recent_memory.iter().take(5) {
                text += &format!("  [{}] {}\n", cat, content);
            }
            parts.push(text.trim_end().to_string());
        }
        if !self.variables.is_empty() {
            let user_vars: Vec<(&String, &String)> = self
                .variables
                .iter()
                .filter(|(_, v)| {
                    !v.starts_with("SynValue(task:")
                        && !v.starts_with("builtin:")
                        && !v.starts_with("task ")
                })
                .collect();
            if !user_vars.is_empty() {
                let mut text = String::from("Visible variables:\n");
                for (k, v) in user_vars.iter().take(15) {
                    text += &format!("  {} = {}\n", k, v);
                }
                parts.push(text.trim_end().to_string());
            }
        }
        if !self.capabilities.is_empty() {
            parts.push(format!("Available capabilities: {}", self.capabilities.join(", ")));
        }
        parts.push(
            "Respond concisely and directly. When choosing between options, respond with ONLY \
             the chosen option. When analyzing, be structured and actionable. Respect all owner \
             rules."
                .to_string(),
        );
        parts.join("\n\n")
    }
}

/// Construye un prompt completo con contexto para una operación del LLM.
pub fn build_contextual_prompt(
    operation: &str,
    data: &HashMap<String, String>,
    context: Option<&LLMContext>,
) -> String {
    let system = context.map(|c| c.build_system_prompt()).unwrap_or_default();
    let g = |k: &str| data.get(k).cloned().unwrap_or_default();

    let prompt = match operation {
        "reason" => {
            let mut p = format!("Reason about: {}\n", g("subject"));
            p += "\nProvide a clear, structured analysis.";
            p
        }
        "decide" => {
            let mut p = format!("Given: {}\n\n", g("given"));
            p += &format!("Choose the best option from: {}\n", g("options"));
            p += "Respond with ONLY the chosen option, nothing else.";
            p
        }
        "analyze" => {
            let mut p = format!("Data to analyze:\n{}\n\n", g("data"));
            p += &format!("Objective: {}\n", g("objective"));
            p += "Provide a concise, actionable analysis.";
            p
        }
        "generate" => {
            let mut p = format!("Generate: {}\n", g("target"));
            let given = g("given");
            if !given.is_empty() {
                p += &format!("\nBased on: {}\n", given);
            }
            p
        }
        _ => format!("{:?}", data),
    };

    if !system.is_empty() {
        format!("{}\n\n---\n\n{}", system, prompt)
    } else {
        prompt
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn system_prompt_includes_context() {
        let mut c = LLMContext::new();
        c.set_intent("Handle support tickets");
        c.rules.push(("hard".to_string(), "max_refund <= 200".to_string()));
        c.variables.insert("customer".to_string(), "Alice".to_string());
        let p = c.build_system_prompt();
        assert!(p.contains("reasoning engine"));
        assert!(p.contains("Handle support tickets"));
        assert!(p.contains("max_refund <= 200"));
        assert!(p.contains("customer = Alice"));
    }

    #[test]
    fn system_prompt_filters_builtins() {
        let mut c = LLMContext::new();
        c.variables.insert("print".to_string(), "builtin:print".to_string());
        c.variables.insert("x".to_string(), "42".to_string());
        let p = c.build_system_prompt();
        assert!(p.contains("x = 42"));
        assert!(!p.contains("builtin:print"));
    }

    #[test]
    fn contextual_prompt_decide() {
        let mut d = HashMap::new();
        d.insert("options".to_string(), "[refund, replace]".to_string());
        d.insert("given".to_string(), "broken item".to_string());
        let p = build_contextual_prompt("decide", &d, None);
        assert!(p.contains("Choose the best option from: [refund, replace]"));
        assert!(p.contains("ONLY the chosen option"));
    }

    #[test]
    fn contextual_prompt_analyze_with_system() {
        let mut c = LLMContext::new();
        c.set_intent("Audit logs");
        let mut d = HashMap::new();
        d.insert("data".to_string(), "logs".to_string());
        d.insert("objective".to_string(), "find anomalies".to_string());
        let p = build_contextual_prompt("analyze", &d, Some(&c));
        assert!(p.contains("Audit logs")); // del system prompt
        assert!(p.contains("Objective: find anomalies"));
        assert!(p.contains("---")); // separador system/prompt
    }
}
